File: macapp/TodayH1/Sources/TodayH1/SyncClient.swift
import Foundation
import CryptoKit

struct PullResponse: Decodable {
    let found: Bool
    let ciphertext: String?
    let iv: String?
    let updated_at: String?
}

struct PushResponse: Decodable {
    let applied: Bool
    let updated_at: String
}

/// Same shape as the JSON object index.html's encryptBlob wraps for the
/// "entries" channel: { version, updatedAt, entries, h1Shortcuts }.
struct EntriesBlob: Codable {
    var version: Int
    var updatedAt: String
    var entries: [String: String]
    var h1Shortcuts: String
}

enum SyncError: Error, LocalizedError {
    case invalidBaseURL
    case network(String)
    case pushRejected

    var errorDescription: String? {
        switch self {
        case .invalidBaseURL: return "API URLが不正です"
        case .network(let m): return "通信エラー: \(m)"
        case .pushRejected: return "サーバー側に新しい変更があるため保存できませんでした"
        }
    }
}

extension Date {
    /// Matches JS's `new Date().toISOString()` exactly (millisecond
    /// precision, trailing "Z"), since the server's last-write-wins
    /// comparison and the web app's own writes both use that format.
    var jsISOString: String {
        let df = DateFormatter()
        df.calendar = Calendar(identifier: .gregorian)
        df.locale = Locale(identifier: "en_US_POSIX")
        df.timeZone = TimeZone(identifier: "UTC")
        df.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
        return df.string(from: self)
    }
}

enum FlexibleISO8601 {
    private static let withFractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
    private static let withoutFractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    /// The server (Python) emits ISO8601 with a "+00:00" offset and
    /// microsecond precision that's omitted when exactly zero — try
    /// both so we never fail to parse a valid response.
    static func date(from s: String) -> Date? {
        withFractional.date(from: s) ?? withoutFractional.date(from: s)
    }
}

final class SyncClient {
    let baseURL: URL
    let syncId: String
    let aesKey: SymmetricKey

    init(baseURLString: String, seedString: String) throws {
        guard let url = URL(string: baseURLString) else { throw SyncError.invalidBaseURL }
        self.baseURL = url
        let seedBytes = try SyncCrypto.seedBytes(from: seedString)
        self.syncId = SyncCrypto.deriveSyncId(seed: seedBytes)
        self.aesKey = SyncCrypto.deriveAesKey(seed: seedBytes)
    }

    static func todayDateKey(now: Date = Date()) -> String {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone.current
        let c = cal.dateComponents([.year, .month, .day], from: now)
        return String(format: "%04d-%02d-%02d", c.year!, c.month!, c.day!)
    }

    private func pull() async throws -> (blob: EntriesBlob, serverUpdatedAt: Date?) {
        var comps = URLComponents(
            url: baseURL.appendingPathComponent("api/pull"),
            resolvingAgainstBaseURL: false
        )!
        comps.queryItems = [URLQueryItem(name: "sync_id", value: syncId)]
        var req = URLRequest(url: comps.url!)
        req.timeoutInterval = 10

        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse, http.statusCode == 200 else {
            throw SyncError.network("pull failed")
        }
        let body = try JSONDecoder().decode(PullResponse.self, from: data)
        guard body.found, let ct = body.ciphertext, let iv = body.iv, let updatedAtStr = body.updated_at else {
            return (EntriesBlob(version: 1, updatedAt: Date().jsISOString, entries: [:], h1Shortcuts: "[]"), nil)
        }
        let plaintext = try SyncCrypto.decrypt(ciphertextB64: ct, ivB64: iv, key: aesKey)
        let blob = try JSONDecoder().decode(EntriesBlob.self, from: plaintext)
        return (blob, FlexibleISO8601.date(from: updatedAtStr))
    }

    private func push(blob: EntriesBlob) async throws -> Bool {
        let plaintext = try JSONEncoder().encode(blob)
        let (ciphertext, iv) = try SyncCrypto.encrypt(plaintext: plaintext, key: aesKey)

        var req = URLRequest(url: baseURL.appendingPathComponent("api/push"))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.timeoutInterval = 10
        let payload: [String: Any] = [
            "sync_id": syncId,
            "ciphertext": ciphertext,
            "iv": iv,
            "updated_at": blob.updatedAt,
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse, http.statusCode == 200 else {
            throw SyncError.network("push failed")
        }
        let body = try JSONDecoder().decode(PushResponse.self, from: data)
        return body.applied
    }

    /// Prepends `blockText` to today's entry, exactly like index.html's
    /// insertHistoryBlockIntoCurrent, then pushes the whole (still
    /// end-to-end-encrypted) blob back. If the server rejects the push
    /// because something else changed the data first (e.g. an edit made
    /// in the browser), re-pull the fresh state and retry — this app is
    /// just one more "device" writing to the same sync blob, so it
    /// follows the same last-write-wins protocol the web app uses
    /// between multiple browsers.
    func prependToday(blockText: String, maxAttempts: Int = 3) async throws {
        var lastError: Error = SyncError.pushRejected
        for _ in 0..<maxAttempts {
            do {
                var (blob, serverUpdatedAt) = try await pull()
                let todayKey = Self.todayDateKey()
                let current = blob.entries[todayKey] ?? "# "
                let trimmed = current.trimmingCharacters(in: .whitespacesAndNewlines)
                let base = (trimmed.isEmpty || trimmed == "#") ? "" : current
                let next = base.isEmpty ? blockText : "\(blockText)\n\(base)"
                blob.entries[todayKey] = next
                blob.version = 1

                let now = Date()
                let bumped = serverUpdatedAt.map { max(now, $0.addingTimeInterval(1)) } ?? now
                blob.updatedAt = bumped.jsISOString

                if try await push(blob: blob) {
                    return
                }
                lastError = SyncError.pushRejected
            } catch {
                lastError = error
            }
        }
        throw lastError
    }
}

File: android/app/src/main/java/com/ihcimen/android/sync/SyncClient.kt
package com.ihcimen.android.sync

import com.ihcimen.android.crypto.SyncCrypto
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.concurrent.TimeUnit

class SyncRejectedException : Exception("サーバー側に新しい変更があるため保存できませんでした")

/**
 * One "entries" sync channel client, talking to the same /api/pull and
 * /api/push endpoints as the web app (index.html) and TodayH1
 * (SyncClient.swift), using the identical crypto and wire format. This
 * class only knows how to pull/push a whole blob or prepend today's block
 * — it has no knowledge of local storage; see SyncManager for that.
 */
class SyncClient(
    baseUrl: String,
    seed: String,
    private val client: OkHttpClient = defaultClient,
) {
    private val apiBaseUrl = baseUrl.trimEnd('/')
    private val seedBytes = SyncCrypto.seedBytes(seed)
    val syncId: String = SyncCrypto.deriveSyncId(seedBytes)
    private val aesKey = SyncCrypto.deriveAesKey(seedBytes)
    private val json = Json { ignoreUnknownKeys = true }

    companion object {
        private val defaultClient = OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .build()
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
        private val ISO_FORMATTER: DateTimeFormatter =
            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneId.of("UTC"))

        /** yyyy-MM-dd in the device's local timezone, matching localStorage's date keys. */
        fun todayDateKey(zone: ZoneId = ZoneId.systemDefault()): String =
            java.time.LocalDate.now(zone).toString()

        /** Matches JS's `new Date().toISOString()` exactly (millisecond precision, trailing "Z"). */
        fun nowJsIso(instant: Instant = Instant.now()): String = ISO_FORMATTER.format(instant)
    }

    suspend fun pull(): Pair<EntriesBlob, Instant?> = withContext(Dispatchers.IO) {
        val request = Request.Builder().url("$apiBaseUrl/api/pull?sync_id=$syncId").get().build()
        val responseBody = execute(request)
        val body = json.decodeFromString(PullResponse.serializer(), responseBody)
        if (!body.found || body.ciphertext == null || body.iv == null || body.updated_at == null) {
            return@withContext EntriesBlob(updatedAt = nowJsIso()) to null
        }
        val plaintext = SyncCrypto.decrypt(body.ciphertext, body.iv, aesKey)
        val blob = json.decodeFromString(EntriesBlob.serializer(), String(plaintext, Charsets.UTF_8))
        val serverUpdatedAt = OffsetDateTime.parse(body.updated_at).toInstant()
        blob to serverUpdatedAt
    }

    suspend fun push(blob: EntriesBlob): Boolean = withContext(Dispatchers.IO) {
        val plaintext = json.encodeToString(EntriesBlob.serializer(), blob).toByteArray(Charsets.UTF_8)
        val (ciphertext, iv) = SyncCrypto.encrypt(plaintext, aesKey)
        val payload = PushRequestBody(syncId, ciphertext, iv, blob.updatedAt)
        val bodyStr = json.encodeToString(PushRequestBody.serializer(), payload)
        val request = Request.Builder()
            .url("$apiBaseUrl/api/push")
            .post(bodyStr.toRequestBody(JSON_MEDIA_TYPE))
            .build()
        val responseBody = execute(request)
        json.decodeFromString(PushResponse.serializer(), responseBody).applied
    }

    /**
     * Mirrors SyncClient.swift's prependToday: pull the freshest server
     * state, prepend blockText to dateKey's entry within that fresh blob,
     * then push. If the push is rejected because another device wrote in
     * between, re-pull and retry — this device is just one more "sync
     * client" writing to the same blob, following the same last-write-wins
     * protocol the web app uses between browsers.
     */
    suspend fun prependToday(blockText: String, dateKey: String = todayDateKey(), maxAttempts: Int = 3) {
        var lastError: Exception = SyncRejectedException()
        repeat(maxAttempts) {
            try {
                val (blob, serverUpdatedAt) = pull()
                val current = blob.entries[dateKey] ?: "# "
                val trimmed = current.trim()
                val base = if (trimmed.isEmpty() || trimmed == "#") "" else current
                val next = if (base.isEmpty()) blockText else "$blockText\n$base"
                val entries = blob.entries.toMutableMap().apply { put(dateKey, next) }

                val now = Instant.now()
                val bumped = serverUpdatedAt?.let { maxOf(now, it.plusMillis(1)) } ?: now
                val nextBlob = blob.copy(version = 1, updatedAt = nowJsIso(bumped), entries = entries)

                if (push(nextBlob)) return
                lastError = SyncRejectedException()
            } catch (e: Exception) {
                lastError = e
            }
        }
        throw lastError
    }

    private fun execute(request: Request): String {
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("HTTP ${response.code}")
            return response.body?.string() ?: throw IOException("empty response body")
        }
    }
}
